### DISCUSSION TITLE
Hide Payload Notify NAT_DETECTION

### DISCUSSION THREAD

**Discussion by AD4956** [strongswan/strongswan#1334] [Q&A]
Hi

The initiator (strongswan client 5.9.8) is currently sending a NAT_DETECTION_SOURCE_IP & NAT_DETECTION_DESTINATION_IP on the first. 
The problem is that the responder (firewall) rejects the connection because it is configured to not detect nat (he doesn't need to detect it since NATT is mandatory).

Is there a solution to this ? I tried to look around in the source code ike_natd.c but I can't find the way to hide this messages.

Thanks

**Comment by tobiasbrunner** [comment]
> The problem is that the responder (firewall) rejects the connection because it is configured to not detect nat (he doesn't need to detect it since NATT is mandatory).

That doesn't make any sense.

**Comment by AD4956** [comment]
So there are no solutions ?

I can't change anything on the firewall.

**Reply by tobiasbrunner** [reply]
> I can't change anything on the firewall.

What firewall is this? What feature is it that blocks this? It really makes no sense whatsoever.

**Reply by AD4956** [reply]
I'm not sure why it doesn't make sense.. .I'm just struggling to get this mobile vpn working !
The firewall is a Stormshield.

Frame captured from a working vpn client with NAT-T Negociation disabled :

`Internet Protocol Version 4, Src: X.X.X.X, Dst: X.X.X.X
User Datagram Protocol, Src Port: 4500, Dst Port: 4500
UDP Encapsulation of IPsec Packets
Internet Security Association and Key Management Protocol
    Initiator SPI: c7d68998b3f68192
    Responder SPI: 0000000000000000
    Next payload: Security Association (33)
    Version: 2.0
    Exchange type: IKE_SA_INIT (34)
    Flags: 0x08 (Initiator, No higher version, Request)
    Message ID: 0x00000000
    Length: 176
    Payload: Security Association (33)
        Next payload: Key Exchange (34)
        0... .... = Critical Bit: Not critical
        .000 0000 = Reserved: 0x00
        Payload length: 40
        Payload: Proposal (2) # 1
            Next payload: NONE / No Next Payload  (0)
            Reserved: 00
            Payload length: 36
            Proposal number: 1
            Protocol ID: IKE (1)
            SPI Size: 0
            Proposal transforms: 3
            Payload: Transform (3)
                Next payload: Transform (3)
                Reserved: 00
                Payload length: 12
                Transform Type: Encryption Algorithm (ENCR) (1)
                Reserved: 00
                Transform ID (ENCR): AES-GCM with a 16 octet ICV (20)
                Transform Attribute (t=14,l=2): Key Length: 256
            Payload: Transform (3)
                Next payload: Transform (3)
                Reserved: 00
                Payload length: 8
                Transform Type: Pseudo-random Function (PRF) (2)
                Reserved: 00
                Transform ID (PRF): PRF_HMAC_SHA2_256 (5)
            Payload: Transform (3)
                Next payload: NONE / No Next Payload  (0)
                Reserved: 00
                Payload length: 8
                Transform Type: Diffie-Hellman Group (D-H) (4)
                Reserved: 00
                Transform ID (D-H): 256-bit random ECP group (19)
    Payload: Key Exchange (34)
        Next payload: Nonce (40)
        0... .... = Critical Bit: Not critical
        .000 0000 = Reserved: 0x00
        Payload length: 72
        DH Group #: 256-bit random ECP group (19)
        Reserved: 0000
        Key Exchange Data: 2983d5099dd216393aa2bcd715d6897198e00a7e9d60cee8a5fa020f8e4c8ce03f90451f…
    Payload: Nonce (40)
        Next payload: NONE / No Next Payload  (0)
        0... .... = Critical Bit: Not critical
        .000 0000 = Reserved: 0x00
        Payload length: 36
        Nonce DATA: def8860b9752be243ac305674dc5d6d8e72820ec80de1108fba5e26665fc164d
`

Frame capture with NAT-T Detection enabled :

`Internet Protocol Version 4, Src: X.X.X.X, Dst: X.X.X.X
User Datagram Protocol, Src Port: 4500, Dst Port: 4500
UDP Encapsulation of IPsec Packets
Internet Security Association and Key Management Protocol
    Initiator SPI: 70793015c70e9f99
    Responder SPI: 0000000000000000
    Next payload: Security Association (33)
    Version: 2.0
    Exchange type: IKE_SA_INIT (34)
    Flags: 0x08 (Initiator, No higher version, Request)
    Message ID: 0x00000000
    Length: 232
    Payload: Security Association (33)
    Payload: Key Exchange (34)
    Payload: Nonce (40)
    Payload: Notify (41) - NAT_DETECTION_SOURCE_IP
    Payload: Notify (41) - NAT_DETECTION_DESTINATION_IP
`

It doesn't go further. On the responder I have this message :

IKE_SA state change: CREATED => CONNECTING
received unexpected nat detection payload
checkin and destroy IKE_SA
IKE_SA state change: CONNECTING => DESTROYING

**Reply by tobiasbrunner** [reply]
> I'm not sure why it doesn't make sense.

Rejecting connections because of some (non-error) notifies is not something that IKEv2 implementations usually do. If a responder doesn't support or doesn't want to use NAT-T it can just omit the NAT-D notifies in its response. There is no need to forcibly terminate the connection. However, you also wrote that NAT detection is explicitly disabled on the responder because it "doesn't need to detect it since NATT is mandatory". That makes even less sense. If NAT-T (i.e. UDP encapsulation) is mandatory, the responder should very much **want** to send NAT-D notifies, just in a way that causes a NAT to be detected and NAT-T to be activated (e.g. strongSwan generates a random `NAT_DETECTION_SOURCE_IP` notify in such a situation). That's the only way to enable NAT-T in an interoperable way.
